WordPress/wordpress-playground
# [Data Liberation] "Fetch from a different URL" button for failed media downloads, Interactivity API support

## Description

Ships:

* Fetch from a different URL" button that puts the user in control of media download failures
* Automatically refreshing the progress details on the wp-admin page via the interactivity API.

### User-driven error handling

When a remote asset cannot be downloaded, most importers either stop or ignore the error. This PR adds a user interaction to make an explicit decision about what should happen next – do we ignore the missing asset? Do we use another file instead?

https://github.com/user-attachments/assets/cea48258-b644-434c-9fb2-1b890c4d86d7


### Auto-Refreshing Import Status

This PR also re-expresses the entire data liberation wp-admin page using the interactivity API, and auto-refreshes the progress:

https://github.com/user-attachments/assets/e093268b-5deb-4bc2-a1d2-e2bb1148e153

A part of https://github.com/WordPress/wordpress-playground/issues/1894

## Technical overview

### User-driven error handling

During the frontloading stage, the `WP_Stream_Importer` exposes all the frontloaded entities to the API consumer. The consumer then creates a post of type `frontloading_placeholder` with an initial status `awaiting_download` for each asset, and updates it with progress information and status (success, failure, skipped) as the import progresses.

The frontloading stage is not finished until all the frontloaded assets have been processed with a non-error outcome. There's a few ways to recover from errors:

* Retry the download – `WP_Stream_Importer` now retries the failed assets URL (via `WP_Retry_Frontloading_Iterator`) before moving on to entities provided by the usual entity source such as a WXR file.
* Changing the downloaded URL – done by the user on the wp-admin page
* Choosing to skip the download – done by the user on the wp-admin page

Sometimes we don't want to require user interactions, e.g. when running the `importWxr` Blueprint step. In those scenarios, we could choose a default error outcome, e.g. "skip failed downloads".

### Auto-refreshing admin page

Two `fetch()` requests running in an infinite loop are:

* Updating the JavaScript interactivity store with the latest import state from the server
* Running the next import step

### Other changes

* Adds `php_userstreamop_read` to the Asyncify list – it crashed the importer in `@wp-playground/cli` running in bun.

## Follow-up work

* Pretty UI transitions. Right now it's all sudden and jerky. We need progress bars, smooth animations, clear visual causality.
* Prevent running the same import in two concurrent requests. This is a serial importer not designed for parallelization. 
* Run each import step in a transaction – either it all worked and we can commit the changes and an updated cursor, or it didn't work and we roll back the last step. Ideally we'll never see a scenario where an entity was processed, but a crash happened before storing the updated cursor and the next run reprocesses the same entity.

## Testing instructions

* Go to the data liberation admin page
* Upload a WXR export file
* Confirm the import processes automatically and doesn't error out
